File: src/test/java/it/marketplace/microservices/config/exception/ServiceExceptionTest.java
package it.marketplace.microservices.config.exception;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class ServiceExceptionTest {
    @Test
    void shouldCreateServiceExceptionWithErrorCodeAndMessage() {
        // Arrange
        ServiceException.ErrorCode code = ServiceException.ErrorCode.PRODUCT_NOT_FOUND;
        String message = "Product not found!";
        // Act
        ServiceException ex = new ServiceException(code, message);
        // Assert
        assertEquals(code, ex.getErrorCode());
        assertEquals(message, ex.getMessage());
    }

    @Test
    void shouldCreateServiceExceptionWithNoArgs() {
        // Act
        ServiceException ex = new ServiceException();
        // Assert
        assertNull(ex.getErrorCode());
        assertNull(ex.getMessage());
    }

    @Test
    void shouldContainAllErrorCodes() {
        // Act & Assert
        assertNotNull(ServiceException.ErrorCode.GENERIC_ERROR);
        assertNotNull(ServiceException.ErrorCode.DATA_ALREADY_PRESENT);
        assertNotNull(ServiceException.ErrorCode.EMAIL_NOT_FOUND);
        assertNotNull(ServiceException.ErrorCode.PRODUCT_NOT_FOUND);
        assertNotNull(ServiceException.ErrorCode.PRODUCT_SUPPLY_EXCEED);
        assertNotNull(ServiceException.ErrorCode.ORDER_NOT_FOUND);
        assertNotNull(ServiceException.ErrorCode.ORDER_EXIST_FOR_USER_FOUND);
        assertNotNull(ServiceException.ErrorCode.PAYMENT_RATE_EXCEED);
    }
}

File: src/test/java/it/marketplace/microservices/config/exception/ServiceExceptionTest.kt
package it.marketplace.microservices.config.exception

import it.marketplace.microservices.common.enums.ErrorCode
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

class ServiceExceptionTest {
    @Test
    fun shouldCreateServiceExceptionWithErrorCodeAndMessage() {
        // Arrange
        val code = ErrorCode.PRODUCT_NOT_FOUND
        val message = "Product not found!"
        // Act
        val ex = ServiceException(code, message)
        // Assert
        assertEquals(code, ex.errorCode)
        assertEquals(message, ex.message)
    }

    @Test
    fun shouldCreateServiceExceptionWithNoArgs() {
        // Act
        val ex = ServiceException(
            ErrorCode.GENERIC_ERROR,
            "An unexpected error occurred."
        )
        // Assert
        assertNull(ex.errorCode)
        assertNull(ex.message)
    }

    @Test
    fun shouldContainAllErrorCodes() {
        // Act & Assert
        assertNotNull(ErrorCode.GENERIC_ERROR)
        assertNotNull(ErrorCode.DATA_ALREADY_PRESENT)
        assertNotNull(ErrorCode.EMAIL_NOT_FOUND)
        assertNotNull(ErrorCode.PRODUCT_NOT_FOUND)
        assertNotNull(ErrorCode.PRODUCT_SUPPLY_EXCEED)
        assertNotNull(ErrorCode.ORDER_NOT_FOUND)
        assertNotNull(ErrorCode.ORDER_EXIST_FOR_USER_FOUND)
        assertNotNull(ErrorCode.PAYMENT_RATE_EXCEED)
    }
}

